src/main/scala/util/Hierarchy.scala

```diff
@@ -0,0 +1,110 @@
+// See LICENSE for license details.
+
+package sifive.blocks.util
+
+import chisel3._
+import chisel3.util._
+
+import freechips.rocketchip.config._
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree.LogicalTreeNode
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.prci._
+import freechips.rocketchip.util.LocationMap
+
+import firrtl.graph._
+
+case object HierarchyKey extends Field[Option[DiGraph[HierarchicalLocation]]](None)
+
+case class DevicesSubsystemParams(
+  name: String,
+  logicalTreeNode: LogicalTreeNode,
+  asyncClockGroupsNode: ClockGroupEphemeralNode)
+
+class DevicesSubsystem(val location: HierarchicalLocation, val ibus: InterruptBusWrapper, params: DevicesSubsystemParams)(implicit p: Parameters)
+  extends LazyModule 
+    with Attachable
+    with HasConfigurableTLNetworkTopology 
+    with CanHaveDevices {
+
+  def devicesSubhierarchies = None
+  def logicalTreeNode = params.logicalTreeNode
+  implicit val asyncClockGroupsNode = params.asyncClockGroupsNode
+
+  lazy val module = new LazyModuleImp(this) {
+    override def desiredName: String = params.name
+  }
+}
+
+trait HasConfigurableHierarchy { this: Attachable =>
+  def location: HierarchicalLocation
+
+  def createHierarchyMap(
+    root: HierarchicalLocation,
+    graph: DiGraph[HierarchicalLocation],
+    context: Attachable): Unit = {
+
+    hierarchyMap += (root -> context)
+
+    // Create and recurse on child hierarchies
+    val edges = graph.getEdges(root)
+    edges.foreach { edge =>
+      val dssParams = DevicesSubsystemParams(
+        name = edge.name,
+        logicalTreeNode = this.logicalTreeNode,
+        asyncClockGroupsNode = this.asyncClockGroupsNode)
+      val dss = context { LazyModule(new DevicesSubsystem(edge, ibus, dssParams)) }
```

Comment:
It seems like this should be some kind of `instantiate` function so we can use all this code to populate hierarchies of different kinds of subsystems? Or do your think `DevicesSubsystem` is the new fundamental building block and is the only class we will want to nest in this way? Maybe we should just figure out how to merge it with `BaseSubsystem` (as a next step.)

Reply:
Yeah, my hope is that DevicesSubsystem is used to create all new/nested subsystems from here on out. In the long term I would like to merge this with BaseSubsystem: I think that will mean turning BaseSubsystem into an instantiation of DevicesSubsystem by migrating all of the subhierarchies of BaseSubsystem to use the devices attachment API